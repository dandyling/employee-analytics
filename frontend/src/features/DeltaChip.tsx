import React from "react";
import { Badge, ColorType } from "../components/badge/Badge";

interface Props {
  previous: number;
  current: number;
}

export const DeltaChip = (props: Props) => {
  const { previous, current } = props;

  const difference = Math.round(getDelta(previous, current));
  const variant = getDeltaVariant(difference);

  return (
    <div className="flex items-center justify-center w-full">
      <Badge variant={variant}>{`${difference}%`}</Badge>
    </div>
  );
};

const getDelta = (current: number, previous: number) => {
  return ((current - previous) / previous) * 100;
};

const getDeltaVariant = (delta: number): ColorType => {
  if (delta > 0) {
    return "success";
  } else if (delta < 0) {
    return "warning";
  } else {
    return "normal";
  }
};
