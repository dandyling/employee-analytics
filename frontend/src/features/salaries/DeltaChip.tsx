import React from "react";
import { Badge, ColorType } from "../../components/badge/Badge";

interface Props {
  previous: number;
  current: number;
}

export const DeltaChip = (props: Props) => {
  const { previous, current } = props;

  const delta = getDelta(previous, current);
  if (!delta) {
    return null;
  }

  const difference = Math.round(delta);
  const variant = getDeltaVariant(difference);

  return (
    <div className="flex items-center justify-center w-full">
      <Badge variant={variant}>{`${difference}%`}</Badge>
    </div>
  );
};

export const getDelta = (previous: number, current: number) => {
  if (previous === 0) {
    return null;
  }
  return ((current - previous) / previous) * 100;
};

export const getDeltaVariant = (delta: number): ColorType => {
  if (delta > 0) {
    return "success";
  } else if (delta < 0) {
    return "warning";
  } else {
    return "normal";
  }
};
