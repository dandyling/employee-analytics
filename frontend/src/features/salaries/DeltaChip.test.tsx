import { render, screen } from "@testing-library/react";
import React from "react";
import { DeltaChip, getDelta, getDeltaVariant } from "./DeltaChip";

test("renders correctly", () => {
  const wrapper = <DeltaChip previous={2} current={1} />;
  expect(wrapper).toMatchSnapshot();
});

test("if it calculate percentage correctly", () => {
  render(<DeltaChip previous={1} current={1.5} />);
  const childElement = screen.getByText(/50%/i);
  expect(childElement).toBeInTheDocument();
});

test("does not renders anything if passed 0", () => {
  const { container } = render(<DeltaChip previous={0} current={1} />);
  expect(container.firstChild).toBeNull();
});

test("getDelta returns null if dividy by zero", () => {
  expect(getDelta(0, 1)).toBeNull();
});

test("returns success color if delta > 0", () => {
  expect(getDeltaVariant(0.1)).toEqual("success");
});

test("returns normal color if delta === 0", () => {
  expect(getDeltaVariant(0)).toEqual("normal");
});

test("returns warning color if delta < 0", () => {
  expect(getDeltaVariant(-0.1)).toEqual("warning");
});
