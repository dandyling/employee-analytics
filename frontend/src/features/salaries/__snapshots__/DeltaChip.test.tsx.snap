// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DeltaChip
  current={1}
  previous={2}
/>
`;
